plugin/documents.py

```diff
@@ -213,13 +212,16 @@ def session_buffers_async(self) -> Generator[SessionBuffer, None, None]:
             yield sv.session_buffer
 
     def on_text_changed_async(self, changes: Iterable[sublime.TextChange]) -> None:
-        self._clear_highlight_regions()
         different, current_region = self._update_stored_region_async()
         if self.view.is_primary():
             for sv in self.session_views_async():
                 sv.on_text_changed_async(changes)
         if not different:
             return
+        if "documentHighlight" not in global_settings.disabled_capabilities:
+            self._clear_highlight_regions()
+            self._when_selection_remains_stable_async(self._do_highlights, current_region,
+                                                      after_ms=self.highlights_debounce_time)
```

Comment:
Although ST fires both `on_text_changed_async` and `on_selection_modified_async` when we type,
There is this fn call to `_update_stored_region_async` in `on_selection_modified_async`,
that fn will for some reason return `different == False`, so we won't go into the if block and do the highlight.

```python
    def on_selection_modified_async(self) -> None:
        different, current_region = self._update_stored_region_async()
        if different:  #  different == False
            self._clear_code_actions_annotation()
            if "documentHighlight" not in global_settings.disabled_capabilities:
                self._when_selection_remains_stable_async(self._do_highlights, current_region,
                                                          after_ms=self.highlights_debounce_time)
```

Reply:
Here is the current behavior(st-4000 branch)

![output](https://user-images.githubusercontent.com/22029477/89729086-90284680-da32-11ea-867c-8c811ecc3b18.gif)
 
Note, when I stop typing,
the documentHighlight is not send, until I move the selection (press left, right)... I am not sure if this behaved like this in the past?

But with my PR, i send the documentHighlight on the `on_text_changed_async` so it will highlight the word when I stop typing

Reply:
Note that doing this from both `on_text_changed_async` and `on_selection_modified_async` will trigger request twice once "selection_remains_stable".

We could use `Debouncer` class to avoid that.

Reply:
> Note that doing this from both on_text_changed_async and on_selection_modified_async will trigger request twice once "selection_remains_stable".

It would make sense that it behaves like that, but that is not how logic currently behaves. 
Currently there wont be two request. 
Only one will be send and that is by on_text_changed_async. 
Because the on_selection_modified_async wont pass the `if different` [check](https://github.com/sublimelsp/LSP/pull/1233/files#diff-cbc07fb64262a8d87f66bb8f5fb32e69R254) .`different` will be `False`.

Reply:
I put two print statements on both `on_text_changed_async` and `on_selection_modified_async`
```python
   def on_text_changed_async(self, changes: Iterable[sublime.TextChange]) -> None:
        different, current_region = self._update_stored_region_async()
        print('on_text_changed_async fired') # here
        if self.view.is_primary():
            for sv in self.session_views_async():
                sv.on_text_changed_async(changes)
        if not different:
            return

        print('on_text_changed_async, debounce request')  # here
        if "documentHighlight" not in global_settings.disabled_capabilities:
            self._clear_highlight_regions()
            self._when_selection_remains_stable_async(self._do_highlights, current_region,
                                                      after_ms=self.highlights_debounce_time)

   def on_selection_modified_async(self) -> None:
        different, current_region = self._update_stored_region_async()
        print('on_selection_modified_async fired')  # here
        if different:
            print('on_selection_modified_async, debounce request')   # here
            self._clear_code_actions_annotation()
            if "documentHighlight" not in global_settings.disabled_capabilities:
                self._when_selection_remains_stable_async(self._do_highlights, current_region,
                                                          after_ms=self.highlights_debounce_time)
```

Logs where I just modify the selection:
```
# pressed left
on_selection_modified_async fired
on_selection_modified_async, debounce request
# pressed left
on_selection_modified_async fired
on_selection_modified_async, debounce request
# pressed left
on_selection_modified_async fired
on_selection_modified_async, debounce request
```
Logs where I type:
```
# typed a
on_text_changed_async fired
on_text_changed_async, debounce request
on_selection_modified_async fired
# typed b
on_text_changed_async fired
on_text_changed_async, debounce request
on_selection_modified_async fired
# typed a
on_text_changed_async fired
on_text_changed_async, debounce request
on_selection_modified_async fired
```
NOTE that on_selection_modified_async, debounce request is not fired when I type

Reply:
OK, good. I haven't checked it myself, just assumed that it would not work correctly.